File: 0486-predict-the-winner/0486-predict-the-winner.cpp
class Solution {
public:  
    int scoreDiff(vector<int>& nums, int i, int j, int turn){
        if(i>=j)    return turn*nums[i];
        
        int x = nums[i] + turn*scoreDiff(nums, i+1, j, -turn);
        int y = nums[j] + turn*scoreDiff(nums, i, j-1, -turn);
        
        return turn * max(x, y);
    }

    bool PredictTheWinner(vector<int>& nums) {
        return scoreDiff(nums, 0, nums.size()-1, 1) >=0;
    }
};

File: 0486-predict-the-winner/0486-predict-the-winner.py
class Solution:
    def playerOneScore(self, nums, left, right, turn=1):
        if left > right:
            return 0

        choice1 = self.playerOneScore(nums, left+1, right, -turn)
        choice2 = self.playerOneScore(nums, left, right-1, -turn)
        
        if turn==1:
            return max(choice1+nums[left], choice2+nums[right])
        else:
            return min(choice1, choice2)
        
    def PredictTheWinner(self, nums: List[int], turn=0) -> bool:
        n = len(nums)
        totalScore = sum(nums)
        p1_score = self.playerOneScore(nums, 0, n-1)
        p2_score = totalScore - p1_score
        
        return p1_score - p2_score >= 0
        
        